### PR TITLE
artifact soups

### LINKED ISSUE #11321: Artifacts should be able to produce chemical soups.
## Description
"OH GOD IT PRODUCES POTASSIUM AND WATER"

They should be able to produce a handful of different elemental (basic) chems in a "soup", likely with disregard to what they mix to so funny explosives can happen.

### PR DESCRIPTION
**Pull request by EmoGarbage404**
## About the PR

Closes #11321 
adds an effect that chooses 3 random chems and mixes em together in a puddle.

**Media**

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- add: Added a new artifact effect which creates a puddle of random chemicals. Let's hope they don't react.

### CODE REVIEW

**Comment by PixelTheKermit** [comment]
Cognazine

**Comment by deltanedas** [comment]
i dont like that its random *every time it activates*. you might get an artifact that makes bicar or something else harmless, then reactivate it and it makes cfl3. would be good if it was saved so you can write down what happened on the node's paper, as a warning for anyone activating it

**Comment by EmoGarbage404** [comment]
It isn't randomized each activation.

**Comment by deltanedas** [comment]
oh i misunderstood the code then, very good addition

**Comment by deltanedas** [comment]
also insert soup sensors joke

File: Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Components/ChemicalPuddleArtifactComponent.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    /// <summary>
+    /// The solution where all the chemicals are stored
+    /// </summary>
+    [DataField("chemicalSolution", required: true), ViewVariables(VVAccess.ReadWrite)]
+    public Solution ChemicalSolution = default!;
+
+    /// <summary>
+    /// The different chemicals that can be spawned by this effect
+    /// </summary>
+    [DataField("possibleChemicals", required: true, customTypeSerializer: typeof(PrototypeIdListSerializer<ReagentPrototype>)), ViewVariables]
```
Comment (DrSmugleaf) [inline Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Components/ChemicalPuddleArtifactComponent.cs:31]: You don't need read-only VV for data fields

```suggestion
    [DataField("possibleChemicals", required: true, customTypeSerializer: typeof(PrototypeIdListSerializer<ReagentPrototype>))]
```

**Review by DrSmugleaf — CHANGES_REQUESTED** [review]

File: Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Systems/ChemicalPuddleArtifactSystem.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+/// <summary>
+/// This handles <see cref="ChemicalPuddleArtifactComponent"/>
+/// </summary>
+public sealed class ChemicalPuddleArtifactSystem : EntitySystem
+{
+    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly ArtifactSystem _artifact = default!;
+    [Dependency] private readonly SpillableSystem _spillable = default!;
+
+    public const string NodeDataChemicalList = "nodeDataSpawnAmount";
```
Comment (moonheart08) [inline Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Systems/ChemicalPuddleArtifactSystem.cs:17]: doc comment if this really must be a const here.

**Review by moonheart08 — APPROVED** [review]

File: Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Components/ChemicalPuddleArtifactComponent.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+using Robust.Shared.Serialization.TypeSerializers.Implementations.Custom.Prototype.List;
+
+namespace Content.Server.Xenoarchaeology.XenoArtifacts.Effects.Components;
+
+/// <summary>
+/// This is used for an artifact that creates a puddle of
+/// random chemicals upon being triggered.
+/// </summary>
+[RegisterComponent]
+public sealed class ChemicalPuddleArtifactComponent : Component
```
Comment (AJCM-git) [inline Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Components/ChemicalPuddleArtifactComponent.cs:14]: access

File: Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Components/ChemicalPuddleArtifactComponent.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    public Solution ChemicalSolution = default!;
+
+    /// <summary>
+    /// The different chemicals that can be spawned by this effect
+    /// </summary>
+    [DataField("possibleChemicals", required: true, customTypeSerializer: typeof(PrototypeIdListSerializer<ReagentPrototype>))]
+    public List<string> PossibleChemicals = default!;
+
+    /// <summary>
+    /// The number of different chemicals that will be mixed.
```
Comment (AJCM-git) [inline Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Components/ChemicalPuddleArtifactComponent.cs:35]: the number of chemicals in the puddle*

File: Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Systems/ChemicalPuddleArtifactSystem.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    /// <inheritdoc/>
+    public override void Initialize()
+    {
+        SubscribeLocalEvent<ChemicalPuddleArtifactComponent, ArtifactActivatedEvent>(OnActivated);
+    }
+
+    private void OnActivated(EntityUid uid, ChemicalPuddleArtifactComponent component, ArtifactActivatedEvent args)
+    {
+        if (!TryComp<ArtifactComponent>(uid, out var artifact))
+            return;
```
Comment (AJCM-git) [inline Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Systems/ChemicalPuddleArtifactSystem.cs:32]: can ArtifactActivatedEvent ever trigger on something without the artifact component?

File: Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Systems/ChemicalPuddleArtifactSystem.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+        if (!_artifact.TryGetNodeData(uid, NodeDataChemicalList, out List<string> chemicalList, artifact))
+        {
+            chemicalList = new();
```
Comment (AJCM-git) [inline Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Systems/ChemicalPuddleArtifactSystem.cs:36]: do you need to re-declare the chemicalList out var here?

**Review by AJCM-git — CHANGES_REQUESTED** [review]

File: Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Systems/ChemicalPuddleArtifactSystem.cs (hunk 5 above)
Comment (EmoGarbage404) [reply Content.Server/Xenoarchaeology/XenoArtifacts/Effects/Systems/ChemicalPuddleArtifactSystem.cs:32]: no, but i'm just getting the component here to save a possible extra resolve.

**Review by DrSmugleaf — APPROVED** [review]

**Review by AJCM-git — APPROVED** [review]

**Pull request merged**